### ISSUE TITLE
100GB of orphan packages & droplets in blobstore

### ISSUE THREAD

**Issue by rkoster** [cloudfoundry/cloud_controller_ng#440]
In one of our client's environments we noticed an unexpected increase in blobstore size. Further investigation revealed their where a lot of orphan packages laying around (I used [this script](https://gist.github.com/rkoster/9a95ee65eb79988dfaae) to find the orphan packages). I also noticed a lot of failed jobs in the worker queue.

I have not yet found the root cause of the issue but was able to reproduce the behaviour on bosh-lite by running the following bash snippet:
```bash
for i in {1..10}; do
    echo "PUSH: $i"
    dd if=/dev/random of=big_file count=102400 bs=1024
    cf push foo &
    sleep 10
    cf delete-space -f bar && cf create-space bar && cf target -s bar
    kill $!
done
```

The sleep statement in the above script is something which needs to be tweaked according to your setup. 

I was able to see packages build up (while there where no apps deployed) by running:
```bash
bosh ssh api_z1/0
sudo su -
watch ls -lah /var/vcap/store/10.244.0.34.xip.io-cc-packages/??/??/
```

I'm aware that the above is not a typical use-case, but it demonstrates packages can become orphaned. Maybe it would make sense to have a scheduled orphan package/droplets cleaner job.

**Comment by Gerg** [comment]
Using your ruby script, we didn't see any orphans on our test environments. We also tried to reproduce this on bosh-lite using the provided bash snippet. Even after playing around with the length of the sleep, we did not create any orphans. 

Have you made any progress on finding the underlying cause? Any additional information or reproduction steps you can provide will help us characterize this issue.

**Comment by rkoster** [comment]
On further investigation I found multiple failed delayed_jobs for `VCAP::CloudController::Jobs::Runtime::AppBitsPacker`. The passed `app_guid`s matched the files a found in `/var/vcap/store/10.244.0.34.xip.io-cc-packages/`.

Below the full stacks trace from `last_error`, in short the application was `nil`. Which happend because it was deleted before the task completed. After a while the tasks get cleaned up but the blobs stay in the blobstore.

Full stack trace:
```
undefined method `allow_ssh' for nil:NilClass
/var/vcap/data/packages/cloud_controller_ng/9b84bcc79ce079128bb3a06c360d465eb83ad0d1.1-d489916692c5925a0ea8111f9fe0d2dcb066b7f7/cloud_controller_ng/app/models/runtime/constraints/enable_ssh_policy.rb:10:in `validate'
/var/vcap/data/packages/cloud_controller_ng/9b84bcc79ce079128bb3a06c360d465eb83ad0d1.1d489916692c5925a0ea8111f9fe0d2dcb066b7f7/cloud_controller_ng
/app/models/runtime/app.rb:122:in `map'
/var/vcap/data/packages/cloud_controller_ng/9b84bcc79ce079128bb3a06c360d465eb83ad0d1.1d489916692c5925a0ea8111f9fe0d2dcb066b7f7cloud_controller_ng/app/models/runtime/app.rb:122:in `validate'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/sequel-4.21.0/lib/sequel/model/base.rb:1999:in `block in _valid?'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/sequel-4.21.0/lib/sequel/model/base.rb:1124:in `around_validation'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby
/2.1.0/gems/sequel-4.21.0/lib/sequel/model/base.rb:1989:in `_valid?'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/sequel-4.21.0/lib/sequel/model/base.rb:1544:in `block in save'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/task.rb:172:in `invoke_with_call_chain'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/task.rb:165:in `invoke'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:150:in `invoke_task'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:106:in `block (2 levels) in top_level'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:106:in `each'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:106:in `block in top_level'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:115:in `run_with_threads'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:100:in `top_level'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:78:in `block in run'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:176:in `standard_exception_handling'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/lib/rake/application.rb:75:in `run'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/gems/rake-10.4.2/bin/rake:33:in `<top (required)>'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/bin/rake:23:in `load'
/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/bundle/ruby/2.1.0/bin/rake:23:in `<main>'
```

**Comment by dieucao** [comment]
@rkoster 
We do think some type of cleaner job makes sense.  
Would you be interested in submitting a PR for this?
We could discuss requirements etc, if so.

**Comment by rkoster** [comment]
Yes I'm interested in submitting a PR. lets discuss requirements.

**Comment by dieucao** [comment]
@zrob the anchor on the CAPI team is on vacation the rest of this week. I'd want to check in with him for any other considerations.

Couple of things to consider:
Ensuring the deletes actually work on nfs and on s3/compatible
Ensuring the cleaner is light weight and doesn't tie up the cc's for long periods of time scanning blobs.

**Comment by dieucao** [comment]
@rkoster 

We think that we should consider generating periodic(weekly? every third day?) background jobs that can cleanup orphaned data. (the frequency should be configurable)
as a first pass we should focus on droplet and package blobs
considerations for a job like this:
never delete something that is not definitely an orphan
it does not stress the blobstore or cc excessively. consider pagination, self rate-limiting, stopping after a limit because next run can get always get more.
v2/v3 objects create blobs in the same store and may use different key schemes, need to reconcile against both
what happens if a key scheme changes? is there a race condition that would cause us to delete a non-orphaned package or droplet?  For example during a rolling deploy when a newer cloud controller uploads using the new scheme and an older cloud controller is running the job.
it is a valid configuration to use the same bucket for multiple blob types. this means an "orphaned" package could really be a droplet, etc.

Let me know if you need more guidance or have questions.

**Comment by rkoster** [comment]
@dieucao thanks, will start working on this and get back when I have more questions.

**Comment by rkoster** [comment]
Just created: https://github.com/cloudfoundry/cloud_controller_ng/pull/461
Which covers the cleanup of orphaned packages.
Next up cleanup of orphaned droplets.

**Comment by rkoster** [comment]
#461 now also includes orphaned droplets cleanup

**Comment by dieucao** [comment]
@rkoster I've moved this PR over to the backlog which hopefully the team will get to for a deeper review in the next couple of weeks.
In the mean time, could you summarize for me the approach that's been taken?

**Comment by dieucao** [comment]
@rkoster Is it okay to close this issue based on our previous discussion about the PR?

**Comment by rkoster** [comment]
Yes lets close it
